### PR TITLE
PicoLibc support as alternative to Newlib(-nano)

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This PR adds support to use [PicoLibc](https://github.com/keith-packard/picolibc/) as libc implementation on RIOT. PicoLibc at the moment uses Newlib as a basis with the stdio implementation from avr-libc strapped on. For the installation of PicoLibc I'd like to refer to the [build guide](https://github.com/keith-packard/picolibc/blob/master/doc/build.md). For RIOT the default build settings should work fine.

Main advantage of PicoLibc is that it shaves approximately 3KB from the flash footprint of a binary. In case of the hello-world example on a samr21-xpro:

|                      |Total   |Total(LTO=1) | libc  | libc(LTO=1) |
| -----------------|--------|-----------------|----------|------------|
| Newlib-nano: |   8.5KB |  7.8KB      |  4.1KB | 4.1KB    |
| PicoLibc:       |5.6KB  |4.9KB           | 1.3KB  | 1.3KB   |

### Testing procedure

Enable with:
```ShellSession
make PICOLIBC=1
```
such as:
```ShellSession
make -C examples/gnrc_networking BOARD=samr21-xpro PICOLIBC=1
```

For now only the Cortex-m and RISC-V MCUs are supported.

### Issues/PRs references

None

Todo:
- [ ] implement syscalls related to VFS
- [ ] Look at Thread-Local Storage integration
- [ ] Test `printf_float`

### CODE REVIEW

**Comment by bergzand** [comment]
Some benchmarks on a samr21-xpro:

|            | Newlib-Nano | PicoLibc |
| -------- | ------------------ | ------------ |
| bench_msg_pingpong | 56336 | 56736 |
| bench_thread_flags_pingpong |  61854 | 62743 |
| bench_thread_yield_pingpong | 86328 | 87589 |
| bench_mutex_pingpong | 61854 | 62336  |
| bench_sched_nop | 183902 | 185322 |

### bench_runtime_coreapis

Picolibc:
```
Runtime of Selected Core API functions

                 nop loop:    104173us  ---   0.104us per call  ---    9599416 calls per sec

             mutex_init():         7us  ---   0.000us per call  ---  1123222089 calls per sec
        mutex lock/unlock:   2375007us  ---   2.375us per call  ---     421051 calls per sec

       thread_flags_set():   1458340us  ---   1.458us per call  ---     685711 calls per sec
     thread_flags_clear():   1541674us  ---   1.541us per call  ---     648645 calls per sec
thread flags set/wait any:   4333340us  ---   4.333us per call  ---     230768 calls per sec
thread flags set/wait all:   3791674us  ---   3.791us per call  ---     263735 calls per sec
thread flags set/wait one:   4354173us  ---   4.354us per call  ---     229664 calls per sec

        msg_try_receive():   2250007us  ---   2.250us per call  ---     444443 calls per sec
              msg_avail():    520840us  ---   0.520us per call  ---    1919975 calls per sec
```

Newlib-Nano:
```
Runtime of Selected Core API functions

                 nop loop:    104173us  ---   0.104us per call  ---    9599416 calls per sec

             mutex_init():         7us  ---   0.000us per call  ---  1123222089 calls per sec
        mutex lock/unlock:   2520840us  ---   2.520us per call  ---     396693 calls per sec

       thread_flags_set():   1520840us  ---   1.520us per call  ---     657531 calls per sec
     thread_flags_clear():   1562507us  ---   1.562us per call  ---     639997 calls per sec
thread flags set/wait any:   4437507us  ---   4.437us per call  ---     225351 calls per sec
thread flags set/wait all:   3895840us  ---   3.895us per call  ---     256684 calls per sec
thread flags set/wait one:   4479174us  ---   4.479us per call  ---     223255 calls per sec

        msg_try_receive():   2312507us  ---   2.312us per call  ---     432431 calls per sec
              msg_avail():    520840us  ---   0.520us per call  ---    1919975 calls per sec
```

**Comment by benpicco** [comment]
Would it be possible to include picolibc in https://github.com/RIOT-OS/toolchains ?

**Comment by kaspar030** [comment]
> Would it be possible to include picolibc in https://github.com/RIOT-OS/toolchains ?

Definitely, if that needs special toolchain support!

**Comment by benpicco** [comment]
Looks like we'll also need thread-local storage - https://github.com/zephyrproject-rtos/zephyr/pull/26545

**Comment by keith-packard** [comment]
> Looks like we'll also need thread-local storage - [zephyrproject-rtos/zephyr#26545](https://github.com/zephyrproject-rtos/zephyr/pull/26545)

You can build picolibc without TLS, but that means all libc calls with persistent state will share that across threads, including errno. Note that picolibc can only support TLS if the C compiler it was built with supports TLS, so make sure TLS is enabled in your compiler build (it used to be disabled in a bunch of toolchain releases for ARM).

**Comment by keith-packard** [comment]
> Looks like we'll also need thread-local storage - [zephyrproject-rtos/zephyr#26545](https://github.com/zephyrproject-rtos/zephyr/pull/26545)

I've got something working here:

https://github.com/keith-packard/RIOT/tree/picolibc_initial

This is based off this PR with some additional fixes added.

**Comment by bergzand** [comment]
Hi @keith-packard 

Those commits are exactly what we need here. Do you mind if I pull them into this PR, of course keeping your commit authorship for those?

**Comment by keith-packard** [comment]
> Hi @keith-packard
> 
> Those commits are exactly what we need here. Do you mind if I pull them into this PR, of course keeping your commit authorship for those?

That would be awesome.

**Comment by bergzand** [comment]
Now includes the commits from @keith-packard

**Comment by benpicco** [comment]
Now this needs a rebase.

**Comment by bergzand** [comment]
> Now this needs a rebase.

Rebased and updated

File: sys/picolibc_syscalls_default/syscalls.c [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+int picolibc_put(char c, FILE *file)
+{
+    (void)file;
+    stdio_write(&c, 1);
+    return 1;
+}
+FILE picolibc_stdio =
+    FDEV_SETUP_STREAM(picolibc_put, picolibc_get, NULL, _FDEV_SETUP_RW);
```
Comment (benpicco) [inline sys/picolibc_syscalls_default/syscalls.c:80]: Will there be a flush function in case of CDC ACM stdio?

File: sys/picolibc_syscalls_default/syscalls.c [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+{
+    return thread_getpid();
+}
+
+int close(int fd)
+{
+    (void) fd;
+    errno = ENODEV;
+    return -1;
+}
```
Comment (benpicco) [inline sys/picolibc_syscalls_default/syscalls.c:114]: What about `open`?

File: sys/picolibc_syscalls_default/syscalls.c [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+/*
+ * All output is directed to stdio_uart, independent of the given file descriptor.
+ * The write call will further block until the byte is actually written to the UART.
+ */
+_READ_WRITE_RETURN_TYPE write(int fd, const void *data, size_t count)
+{
+    (void) fd;
+    return stdio_write(data, count);
+}
```
Comment (benpicco) [inline sys/picolibc_syscalls_default/syscalls.c:96]: Can't we just call `vfs_write()` here?

File: cpu/fe310/Makefile.include [hunk 4]
```diff
@@ -1,3 +1,12 @@
+ifneq (,$(filter picolibc,$(USEMODULE)))
+USEMODULE += newlib_syscalls_fe310
+endif
+
+USEMODULE += sifive_drivers_fe310
+
+USEMODULE += periph
+USEMODULE += periph_pm
```
[the remark this answers is not in the document]
Comment (bergzand) [reply cpu/fe310/Makefile.include:8]: Yup

File: makefiles/libc/picolibc.mk [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+endif
+
+ifeq (1,$(USE_PICOLIBC))
+  export LINKFLAGS += -specs=picolibc.specs
+endif
+
+export LINKFLAGS += -lc
+
+ifeq (1,$(USE_PICOLIBC))
+  PICOLIBC_INCLUDE_DIR ?= /usr/local/picolibc/$(TARGET_ARCH)/include
```
[the remark this answers is not in the document]
Comment (bergzand) [reply makefiles/libc/picolibc.mk:19]: I think we can do something similar to [`NEWLIB_INCLUDE_DIR`](https://github.com/RIOT-OS/RIOT/blob/6ffac52b1fb6331744466d0a0d134322a1e272d0/makefiles/libc/newlib.mk) here

File: sys/picolibc_syscalls_default/syscalls.c [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    return 1;
+}
+FILE picolibc_stdio =
+    FDEV_SETUP_STREAM(picolibc_put, picolibc_get, NULL, _FDEV_SETUP_RW);
+
+FILE *const __iob[] = {
+    &picolibc_stdio,
+    &picolibc_stdio,
+    &picolibc_stdio,
+};
```
[the remark this answers is not in the document]
Comment (keith-packard) [reply sys/picolibc_syscalls_default/syscalls.c:86]: Yup. You can share the same FILE between all three pointers.

File: sys/picolibc_syscalls_default/syscalls.c (hunk 1 above)
Comment (keith-packard) [reply sys/picolibc_syscalls_default/syscalls.c:80]: Yes, what's the API for that? I didn't see one in the newlib code.

File: core/include/thread.h [hunk 7]
```diff
@@ -176,6 +176,9 @@ struct _thread {
     const char *name;               /**< thread's name                  */
     int stack_size;                 /**< thread's stack size            */
 #endif
+#ifdef PICOLIBC_TLS
```
[the remark this answers is not in the document]
Comment (keith-packard) [reply core/include/thread.h:179]: PICOLIBC_TLS is defined in picolibc.h when picolibc is built with TLS support. That depends on both TLS support in picolibc for the target architecture and TLS support in the compiler. I'm not sure how you'd like to handle this case.

File: sys/picolibc_syscalls_default/syscalls.c (hunk 3 above)
Comment (keith-packard) [reply sys/picolibc_syscalls_default/syscalls.c:96]: > `picolibc_syscalls_default/syscalls.c` and `newlib_syscalls_default/syscalls.c` can probably share some code, e.g. no need to implement `_sbrk_r` twice.

A couple of thoughts here:

 * picolibc provides a version of sbrk that uses all remaining RAM, if you want to use that.
 * picolibc doesn't need _sbrk_r, it just wants regular sbrk. You could save a bit of space by refactoring to eliminate the extra function.

Comment (benpicco) [reply sys/picolibc_syscalls_default/syscalls.c:96]: Will picolibc have problems if the chunks returned by `sbrk` are not continuous?
E.g. some MCUs have non-continuous blocks of RAM, unusually intended for a special function, but if that's not used we just add that to the heap.
We noticed that newlib does something clever internally that breaks if there are gaps in the addresses `_sbrk_r` returns.
With newlib-nano it would just work, so I guess picolib wouldn't do anything special here either.

Comment (keith-packard) [reply sys/picolibc_syscalls_default/syscalls.c:96]: Picolibc is just using the nano-mallocr from newlib, which looks like it will be fine with non-contiguous blocks returned from sbrk. The more complicated regular malloc in newlib might well have troubles. I think the only issue with non-continuous blocks is that mallinfo will return bogus information.

File: sys/picolibc_syscalls_default/syscalls.c [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+ */
+
+#include <errno.h>
+#include <stdio.h>
+
+#include "log.h"
+#include "periph/pm.h"
+#include "stdio_base.h"
+
+uint8_t sbrk[1];
```
[the remark this answers is not in the document]
Comment (bergzand) [reply sys/picolibc_syscalls_default/syscalls.c:30]: leftover from a long time ago I guess.

**Comment by bergzand** [comment]
Compiling and testing on the nrf52dk these are resulting failures:
<details> <summary>failures</summary>

```
ERROR:nrf52dk:Tests failed: 58                                                           
Failures during compilation:                                                             
- [examples/dtls-echo](examples/dtls-echo/compilation.failed)                            
- [examples/dtls-sock](examples/dtls-sock/compilation.failed)                            
- [examples/dtls-wolfssl](examples/dtls-wolfssl/compilation.failed)                      
- [examples/filesystem](examples/filesystem/compilation.failed)                          
- [examples/javascript](examples/javascript/compilation.failed)                          
- [examples/lua_REPL](examples/lua_REPL/compilation.failed)                              
- [examples/lua_basic](examples/lua_basic/compilation.failed)                            
- [examples/micropython](examples/micropython/compilation.failed)                                                                                                                  
- [examples/riot_and_cpp](examples/riot_and_cpp/compilation.failed)                      
- [tests/cpp11_condition_variable](tests/cpp11_condition_variable/compilation.failed)                                                                                              
- [tests/cpp11_mutex](tests/cpp11_mutex/compilation.failed)                                                                                                                        
- [tests/cpp11_thread](tests/cpp11_thread/compilation.failed)                            
- [tests/cpp_ctors](tests/cpp_ctors/compilation.failed)                                                                                                                            
- [tests/cpp_exclude](tests/cpp_exclude/compilation.failed)                                                                                                                        
- [tests/cpp_ext](tests/cpp_ext/compilation.failed)                                      
- [tests/mcuboot](tests/mcuboot/compilation.failed)                                      
- [tests/pkg_emlearn](tests/pkg_emlearn/compilation.failed)                                                                                                                        
- [tests/pkg_flatbuffers](tests/pkg_flatbuffers/compilation.failed)                      
- [tests/pkg_libcose](tests/pkg_libcose/compilation.failed)                              
- [tests/pkg_nanopb](tests/pkg_nanopb/compilation.failed)                                
- [tests/pkg_relic](tests/pkg_relic/compilation.failed)                                  
- [tests/pkg_tensorflow-lite](tests/pkg_tensorflow-lite/compilation.failed)              
- [tests/pkg_tinycbor](tests/pkg_tinycbor/compilation.failed)                            
- [tests/pkg_tinydtls_sock_async](tests/pkg_tinydtls_sock_async/compilation.failed)      
- [tests/pkg_ubasic](tests/pkg_ubasic/compilation.failed)                                
- [tests/pkg_utensor](tests/pkg_utensor/compilation.failed)                              
- [tests/pkg_wolfssl](tests/pkg_wolfssl/compilation.failed)                              
                                                                                         
Failures during test:                                                                                                                                                              
- [examples/suit_update](examples/suit_update/test.failed)             
- [tests/emcute](tests/emcute/test.failed)                                                                                                                                         
- [tests/gnrc_dhcpv6_client](tests/gnrc_dhcpv6_client/test.failed)      
- [tests/gnrc_dhcpv6_client_6lbr](tests/gnrc_dhcpv6_client_6lbr/test.failed)             
- [tests/gnrc_ipv6_ext](tests/gnrc_ipv6_ext/test.failed)                                                                                                                           
- [tests/gnrc_ipv6_ext_frag](tests/gnrc_ipv6_ext_frag/test.failed)                                                                                                                 
- [tests/gnrc_ipv6_ext_opt](tests/gnrc_ipv6_ext_opt/test.failed)                                                                                                                   
- [tests/gnrc_ipv6_nib_dns](tests/gnrc_ipv6_nib_dns/test.failed)                                                                                                                   
- [tests/gnrc_rpl_srh](tests/gnrc_rpl_srh/test.failed)                                   
- [tests/gnrc_sock_dns](tests/gnrc_sock_dns/test.failed)                                 
- [tests/gnrc_tcp](tests/gnrc_tcp/test.failed)                                           
- [tests/malloc](tests/malloc/test.failed)                                               
- [tests/nordic_softdevice](tests/nordic_softdevice/test.failed)                         
- [tests/periph_timer_short_relative_set](tests/periph_timer_short_relative_set/test.failed)                                                                                       
- [tests/pkg_fatfs_vfs](tests/pkg_fatfs_vfs/test.failed)           
- [tests/pkg_hacl](tests/pkg_hacl/test.failed)                                           
- [tests/pkg_libfixmath_unittests](tests/pkg_libfixmath_unittests/test.failed)
- [tests/pkg_littlefs](tests/pkg_littlefs/test.failed)                                   
- [tests/pkg_littlefs2](tests/pkg_littlefs2/test.failed)           
- [tests/pkg_semtech-loramac](tests/pkg_semtech-loramac/test.failed)       
- [tests/pkg_tiny-asn1](tests/pkg_tiny-asn1/test.failed)                                 
- [tests/pthread](tests/pthread/test.failed)                 
- [tests/pthread_barrier](tests/pthread_barrier/test.failed)                             
- [tests/pthread_cleanup](tests/pthread_cleanup/test.failed)
- [tests/pthread_cooperation](tests/pthread_cooperation/test.failed)                     
- [tests/pthread_flood](tests/pthread_flood/test.failed)                                 
- [tests/pthread_tls](tests/pthread_tls/test.failed)                      
- [tests/riotboot](tests/riotboot/test.failed)                                           
- [tests/shell](tests/shell/test.failed)                                                 
- [tests/struct_tm_utility](tests/struct_tm_utility/test.failed)                         
- [tests/xtimer_now_irq](tests/xtimer_now_irq/test.failed)       
```

</details>

Most of the compile-time failures are related to C++ issues, `::vsscanf` being undeclared is the first issue I found so far.

File: sys/picolibc_syscalls_default/syscalls.c [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+#else
+/**
+ * @brief manage the heap
+ */
+extern char _heap_start;                 /* start of the heap */
+extern char _heap_end;                 /* end of the heap */
+
+char * __heap_start = &_heap_start;
+char * __heap_end = &_heap_end;
+#endif
```
Comment (bergzand) [inline sys/picolibc_syscalls_default/syscalls.c:48]: There is probably a better and cleaner way to glue this together.

Comment (benpicco) [reply sys/picolibc_syscalls_default/syscalls.c:48]: I still think we should just use the same names everywhere #12328

Comment (bergzand) [reply sys/picolibc_syscalls_default/syscalls.c:48]: That would at least simplify things here

File: makefiles/libc/picolibc.mk (hunk 5 above)
[the remark this answers is not in the document]
Comment (keith-packard) [reply makefiles/libc/picolibc.mk:19]: If you use --specs=picolibc.specs, then you don't need to have this definition as the directory will get included automatically. picolibc.specs is installed into the cross-compiler root directory so that you don't need a path for that.

File: sys/picolibc_syscalls_default/syscalls.c (hunk 2 above)
Comment (keith-packard) [reply sys/picolibc_syscalls_default/syscalls.c:132]: I'm unclear what this code is even for -- picolibc doesn't need any of these functions. If they aren't otherwise needed, they should just get removed.

File: makefiles/libc/picolibc.mk (hunk 5 above)
[the remark this answers is not in the document]
Comment (bergzand) [reply makefiles/libc/picolibc.mk]: Added the `-specs=picolibc.specs` to the `CFLAGS`, thanks for the hint

File: core/include/thread.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (aabadie) [reply core/include/thread.h:180]: Ok thanks for clarifying. Let's keep it like this then.

File: sys/picolibc_syscalls_default/syscalls.c (hunk 9 above)
Comment (aabadie) [reply sys/picolibc_syscalls_default/syscalls.c:48]: #14455 is merged so should be simpler now.

Comment (bergzand) [reply sys/picolibc_syscalls_default/syscalls.c]: Rebased and simplified!

File: sys/picolibc_syscalls_default/syscalls.c (hunk 1 above)
Comment (keith-packard) [reply sys/picolibc_syscalls_default/syscalls.c:87]: Ok, I reviewed the CDC ACM implementation and that doesn't have any buffering internally; each call to stdio_write causes a separate USB transaction. So, the current code will 'work', although output will be slow. I've added a buffer for output, and also a pile of additional picolibc bits at

 https://github.com/keith-packard/RIOT/tree/pr/libc/picolibc_initial

File: cpu/fe310/start.S [hunk 10]
```diff
@@ -43,8 +43,6 @@ _start:
 2:
 
     /* Call global constructors */
-    la a0, __libc_fini_array
```
Comment (aabadie) [inline cpu/fe310/start.S:46]: I'm wondering if this shouldn't be in its own PR. Using just this change, it saves some bytes on ROM compared to master. And I didn't see any functional regression but I haven't tested in depth (no compile_and_test, just hello-world and default apps).

Comment (keith-packard) [reply cpu/fe310/start.S:46]: It saves memory on all systems, but it's required for picolibc as picolibc has per-thread destructors and this is executed out of thread context (hence there isn't any TLS available here). I guess I'd like to know what global destructors might be ever be registered and if they are expected to be called when exit() is called?

**Comment by bergzand** [comment]
`tests/malloc` and through that `tests/pthread*` are now working on ARM.

The heap start and end were not properly defined. I'm now using `--defsym=__heap_start=_sheap` to glue the different symbol names together. Definitely easier than the previous way.

**Comment by kaspar030** [comment]
pls rebase!

**Comment by bergzand** [comment]
> pls rebase!

Rebased!

**Comment by bergzand** [comment]
With GCC there is still an issue with CPP code:

```
koen@morgen ~/dev/RIOT-picolibc $ PICOLIBC=1 make -C tests/cpp_ext -j BOARD=samr21-xpro
make: Entering directory '/home/koen/dev/RIOT-picolibc/tests/cpp_ext'
Building application "tests_cpp_ext" for "samr21-xpro" with MCU "samd21".

"make" -C /home/koen/dev/RIOT-picolibc/boards/samr21-xpro
"make" -C /home/koen/dev/RIOT-picolibc/core
"make" -C /home/koen/dev/RIOT-picolibc/cpu/samd21
"make" -C /home/koen/dev/RIOT-picolibc/drivers
"make" -C /home/koen/dev/RIOT-picolibc/sys
"make" -C /home/koen/dev/RIOT-picolibc/tests/cpp_ext/module
"make" -C /home/koen/dev/RIOT-picolibc/drivers/periph_common
"make" -C /home/koen/dev/RIOT-picolibc/sys/auto_init
"make" -C /home/koen/dev/RIOT-picolibc/sys/isrpipe
"make" -C /home/koen/dev/RIOT-picolibc/sys/picolibc_syscalls_default
"make" -C /home/koen/dev/RIOT-picolibc/sys/pm_layered
"make" -C /home/koen/dev/RIOT-picolibc/sys/stdio_uart
"make" -C /home/koen/dev/RIOT-picolibc/sys/test_utils/interactive_sync
"make" -C /home/koen/dev/RIOT-picolibc/cpu/cortexm_common
"make" -C /home/koen/dev/RIOT-picolibc/cpu/sam0_common
"make" -C /home/koen/dev/RIOT-picolibc/cpu/samd21/periph
"make" -C /home/koen/dev/RIOT-picolibc/sys/tsrb
"make" -C /home/koen/dev/RIOT-picolibc/cpu/sam0_common/periph
"make" -C /home/koen/dev/RIOT-picolibc/cpu/cortexm_common/periph
/opt/gcc-arm-none-eabi-9-2019-q4-major/bin/../lib/gcc/arm-none-eabi/9.2.1/../../../../arm-none-eabi/bin/ld: /opt/gcc-arm-none-eabi-9-2019-q4-major/bin/../lib/gcc/arm-none-eabi/9.2.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp/libstdc++.a(vterminate.o): in function `__gnu_cxx::__verbose_terminate_handler()':
vterminate.cc:(.text._ZN9__gnu_cxx27__verbose_terminate_handlerEv+0xfc): undefined reference to `_impure_ptr'
collect2: error: ld returned 1 exit status
make: *** [/home/koen/dev/RIOT-picolibc/tests/cpp_ext/../../Makefile.include:574: /home/koen/dev/RIOT-picolibc/tests/cpp_ext/bin/samr21-xpro/tests_cpp_ext.elf] Error 1
make: Leaving directory '/home/koen/dev/RIOT-picolibc/tests/cpp_ext'
```

**Comment by keith-packard** [comment]
> With GCC there is still an issue with CPP code:

> vterminate.cc:(.text._ZN9__gnu_cxx27__verbose_terminate_handlerEv+0xfc): undefined reference to `_impure_ptr'

That's because your libstdc++ is built against newlib instead of picolibc and ends up referencing symbols found in newlib but not in picolibc. I've spent a bit of time adjusting the libstdc++ code that comes with gcc to make it work with picolibc, but that will still require building it correctly and using the right library. The whole libstdc++ issue is a mess -- it should not be included with GCC as that creates circular build dependencies.

**Comment by benpicco** [comment]
So when using picolib, we could still provide the `cpp` feature, but not the `libstdcpp` one. (#14503)
I'd say that's good enough.

File: sys/picolibc_syscalls_default/syscalls.c (hunk 1 above)
Comment (benpicco) [reply sys/picolibc_syscalls_default/syscalls.c:79]: Want to push that pile of additional picolibc bits again? :D 
Or do you want to pull them @bergzand ?

I kind of missed this comment here, I see you hooked up VFS too now.
For the buffering, I wonder if CDC ACM should handle this, but that can be changed later.

**Comment by keith-packard** [comment]
> So when using picolib, we could still provide the `cpp` feature, but not the `libstdcpp` one. (#14503)
> I'd say that's good enough.

Sounds about right. I've got patches for GCC which build libstdc++ against picolibc; perhaps I should start building the Debian embedded toolchains to use those, which would enable libstdc++ with picolibc.

File: sys/picolibc_syscalls_default/syscalls.c (hunk 1 above)
Comment (keith-packard) [reply sys/picolibc_syscalls_default/syscalls.c:79]: I've rebased my changes on top of @bergzand's current branch and pushed them to my branch again.

https://github.com/keith-packard/RIOT/tree/pr/libc/picolibc_initial

I also added a config option so that by default the integer-only printf/scanf code is use, but the full float code can be selected.

Comment (keith-packard) [reply sys/picolibc_syscalls_default/syscalls.c:79]: As for the CDC ACM buffering, it would be nice to include that in the underlying implementation, but that would require exposing a flush function that all users called as needed.

Comment (bergzand) [reply sys/picolibc_syscalls_default/syscalls.c:80]: > I've rebased my changes on top of @bergzand's current branch and pushed them to my branch again.

And pulled them in.

Comment (keith-packard) [reply sys/picolibc_syscalls_default/syscalls.c:80]: I note that you left out the stdout buffering code. That makes USB performance pretty poor; it might be nice to add that back in and plan on removing it if/when the underlying ACM code adds buffering. I'm not really sure it should though -- you'd have to modify all users of that code to know about the buffering so that they could flush when needed.

**Comment by bergzand** [comment]
Closing this one
